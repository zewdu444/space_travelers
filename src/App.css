.nav-list li a.active {
    border-bottom: 4px solid salmon;
  }