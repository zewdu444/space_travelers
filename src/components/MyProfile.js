import React from 'react';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import ListGroup from 'react-bootstrap/ListGroup';
import { useSelector } from 'react-redux';

function MyProfile() {
  const missions = useSelector((state) => state.missions.missionstore);
  return (
    <div className="px-2 pr-2">
      <Row>
        <Col
          sm={12}
          lg={6}
          md={6}
          className="ps-lg-5"
        >
          <h2>My Missions</h2>
          <ListGroup>
            {missions.filter((mission) => mission.reserved === true).map((mission) => (
              <ListGroup.Item key={mission.mission_id}>
                {mission.mission_name}
              </ListGroup.Item>
            ))}
          </ListGroup>
        </Col>
        <Col
          sm={12}
          lg={6}
          md={6}
          className="pe-lg-5"
        >
          <h2>My Rockets</h2>
          <ListGroup>
            <ListGroup.Item>Cras justo odio</ListGroup.Item>
            <ListGroup.Item>Dapibus ac facilisis in</ListGroup.Item>
            <ListGroup.Item>Morbi leo risus</ListGroup.Item>
            <ListGroup.Item>Porta ac consectetur ac</ListGroup.Item>
            <ListGroup.Item>Vestibulum at eros</ListGroup.Item>
          </ListGroup>

        </Col>

      </Row>
    </div>
  );
}

export default MyProfile;
