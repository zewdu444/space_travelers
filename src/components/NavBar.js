import React from 'react';
import { Link } from 'react-router-dom';
import Nav from 'react-bootstrap/Nav';
import Navbar from 'react-bootstrap/Navbar';
import logo from '../asset/planet.png';

function NavBar() {
  return (
    <div>
      <Navbar bg="white" expand="lg" className="px-5">
        <Navbar.Brand to="/" as={Link} className="fs-2 fw-medium">
          <img alt="planet log" src={logo} width={55} height={55} />
          {'  '}
          Space Traveler&#39;s Hub
        </Navbar.Brand>
        <Navbar.Toggle aria-controls="basic-navbar-nav " />
        <Navbar.Collapse
          id="basic-navbar-nav"
          className="justify-content-end"
        >
          <Nav className="text-primary fs-5">
            <Nav.Link className="text-primary pe-5" to="/" as={Link}>Rockets</Nav.Link>
            <Nav.Link className="text-primary" to="/missions" as={Link}>Missions</Nav.Link>
            <div className="d-none d-md-none d-lg-block mt-3 mb-3 vr" />
            <Nav.Link className="text-primary" to="/myprofile" as={Link}>My profile</Nav.Link>
          </Nav>
        </Navbar.Collapse>
      </Navbar>
      <hr />
    </div>
  );
}

export default NavBar;
