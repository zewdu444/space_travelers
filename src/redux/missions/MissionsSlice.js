import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

const BASE_URL = 'https://api.spacexdata.com/v3/missions';

const initialState = {
  missionstore: [],
  status: 'idle',
  error: null,
};
export const fetchMissions = createAsyncThunk('missions/fetchMissions', async () => {
  try {
    const response = await axios.get(BASE_URL);
    return response.data;
  } catch (error) {
    return error.message;
  }
});
export const missionsSlice = createSlice({
  name: 'missions',
  initialState,
  reducers: {
    joinMission: (state, payload) => ({
      ...state,
      missionstore: state.missionstore.map((mission) => {
        if (mission.mission_id === payload.payload) {
          if (mission.reserved === true) {
            return { ...mission, reserved: false };
          }
          return { ...mission, reserved: true };
        }
        return mission;
      }),
    }),
  },
  extraReducers(builder) {
    builder.addCase(fetchMissions.pending, (state) => ({
      ...state,
      status: 'loading',
    }))
      .addCase(fetchMissions.fulfilled, (state, action) => ({
        ...state,
        missionstore: action.payload.map((mission) => ({

          mission_id: mission.mission_id,
          mission_name: mission.mission_name,
          description: mission.description,
        })),
        status: 'loaded',
      })).addCase(fetchMissions.rejected, (state, action) => ({
        ...state,
        status: 'failed',
        error: [...state.error, action.error.message],
      }));
  },

});
export const { joinMission } = missionsSlice.actions;
export default missionsSlice.reducer;
