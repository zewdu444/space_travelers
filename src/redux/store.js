import { configureStore } from '@reduxjs/toolkit';
import { createLogger } from 'redux-logger';
import missionsReducer from './missions/missionsSlice';
import rocketReducer from './rockets/RocketsSlice';

const logger = createLogger({
  // Options for the logger can be passed here
  // Show only the actions in the console
  predicate: (getState, action) => action.type !== 'SOME_ACTION',
  // Collapse the logs by default
  collapsed: true,
});

const store = configureStore({
  reducer: {
    rocckets: rocketReducer,
    missions: missionsReducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(logger),
});

export default store;
