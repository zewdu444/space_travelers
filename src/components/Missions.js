import React, { useEffect } from 'react';
import Table from 'react-bootstrap/Table';
import Button from 'react-bootstrap/Button';
import { useDispatch, useSelector } from 'react-redux';
import Badge from 'react-bootstrap/Badge';
import { fetchMissions, joinMission } from '../redux/missions/missionsSlice';

function Missions() {
  const dispatch = useDispatch();
  const status = useSelector((state) => state.missions.status);
  const missions = useSelector((state) => state.missions.missionstore);
  useEffect(() => {
    if (status === 'idle') {
      dispatch(fetchMissions());
    }
  }, [status, dispatch]);

  const missionHandler = (e) => {
    dispatch(joinMission(e.target.id));
  };
  return (
    <div className="px-5 pr-5">
      <Table striped bordered hover>
        <thead>
          <tr>
            <th>Mission</th>
            <th>Description</th>
            <th>Status</th>
            <th>{}</th>
          </tr>
        </thead>
        <tbody>
          {
          missions.map((mission) => (
            <tr key={mission.mission_id}>
              <td className="fw-bold " style={{ width: '10%' }}>{mission.mission_name}</td>
              <td style={{ width: '45%' }}>
                {mission.description}
              </td>
              <td className="text-center align-middle" style={{ width: '5%' }}>
                <Badge bg={mission.reserved ? ('info') : ('secondary')}>
                  {mission.reserved ? ('Active Member') : ('NOT A MEMBER')}
                </Badge>
              </td>
              <td className="text-center align-middle" style={{ width: '10%' }}>
                <Button
                  id={mission.mission_id}
                  variant={mission.reserved ? ('outline-danger') : ('outline-secondary')}
                  size="md"
                  onClick={missionHandler}
                >
                  {mission.reserved ? ('Leave Misson') : ('Join Misson')}
                </Button>
              </td>
            </tr>
          ))
        }

        </tbody>
      </Table>

    </div>
  );
}

export default Missions;
