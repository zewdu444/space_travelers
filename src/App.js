import { BrowserRouter, Route, Routes } from 'react-router-dom';
import NavBar from './components/NavBar';
import Missions from './components/Missions';
import Rockets from './components/Rockets';
import Myprofile from './components/MyProfile';

function App() {
  return (
    <div className="App">
      <BrowserRouter>
        <NavBar />
        <Routes>
          <Route exact path="/" element={<Rockets />} />
          <Route path="/missions" element={<Missions />} />
          <Route path="/myprofile" element={<Myprofile />} />
        </Routes>
      </BrowserRouter>
    </div>
  );
}

export default App;
