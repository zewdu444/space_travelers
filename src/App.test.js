import React from 'react';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import axios from 'axios';
import NavBar from './components/NavBar';
import Missions from './components/Missions';
import store from './redux/store';
import MyProfile from './components/MyProfile';
import App from './App';
import { fetchMissions, joinMission } from './redux/missions/MissionsSlice';

it('App renders correctly', () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <App />
      </Provider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('NavBar page renders correctly', () => {
  const tree = renderer
    .create(
      <BrowserRouter>
        <NavBar />
      </BrowserRouter>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('Missons page renders correctly', async () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <Missions />
      </Provider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('MyProfile page renders correctly', () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <MyProfile />
      </Provider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

// mission slicer and reducer check

describe('Mission redux state tests', () => {
  it('Should initially set Missions store to an empty Array', () => {
    const state = store.getState().missions;
    expect(state.missionstore).toEqual([]);
  });

  it('should Join Mission payload send correct', () => {
    const expectedPayload = { payload: '9D1B7E0', type: 'missions/joinMission' };
    const actualPayload = joinMission('9D1B7E0');
    expect(actualPayload).toEqual(expectedPayload);
  });

  it('Mission fetch data from API', async () => {
    const url = 'https://api.spacexdata.com/v3/missions';
    const axiosSpy = jest.spyOn(axios, 'get');
    jest.setTimeout(90000);
    const dispatchSpy = jest.fn();

    await fetchMissions(url)(dispatchSpy);

    expect(axiosSpy).toHaveBeenCalledWith(url);
  });
});
