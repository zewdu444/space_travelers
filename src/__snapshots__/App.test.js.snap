// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<div
  className="App"
>
  <div>
    <nav
      className="px-5 navbar navbar-expand-lg navbar-light bg-white"
    >
      <a
        className="fs-2 fw-medium navbar-brand"
        href="/"
        onClick={[Function]}
      >
        <img
          alt="planet log"
          height={55}
          src="planet.png"
          width={55}
        />
          
        Space Traveler's Hub
      </a>
      <button
        aria-controls="basic-navbar-nav "
        aria-label="Toggle navigation"
        className="navbar-toggler collapsed"
        onClick={[Function]}
        type="button"
      >
        <span
          className="navbar-toggler-icon"
        />
      </button>
      <div
        aria-expanded={null}
        className="justify-content-end navbar-collapse collapse"
        id="basic-navbar-nav"
      >
        <div
          className="text-primary fs-5 navbar-nav"
          onKeyDown={[Function]}
        >
          <a
            className="text-primary pe-5 nav-link"
            data-rr-ui-event-key={null}
            disabled={false}
            href="/"
            onClick={[Function]}
          >
            Rockets
          </a>
          <a
            className="text-primary nav-link"
            data-rr-ui-event-key={null}
            disabled={false}
            href="/missions"
            onClick={[Function]}
          >
            Missions
          </a>
          <div
            className="d-none d-md-none d-lg-block mt-3 mb-3 vr"
          />
          <a
            className="text-primary nav-link"
            data-rr-ui-event-key={null}
            disabled={false}
            href="/myprofile"
            onClick={[Function]}
          >
            My profile
          </a>
        </div>
      </div>
    </nav>
    <hr />
  </div>
</div>
`;

exports[`Missons page renders correctly 1`] = `
<div
  className="px-5 pr-5"
>
  <table
    className="table table-striped table-bordered table-hover"
  >
    <thead>
      <tr>
        <th>
          Mission
        </th>
        <th>
          Description
        </th>
        <th>
          Status
        </th>
        <th />
      </tr>
    </thead>
    <tbody />
  </table>
</div>
`;

exports[`MyProfile page renders correctly 1`] = `
<div
  className="px-2 pr-2"
>
  <div
    className="row"
  >
    <div
      className="ps-lg-5 col-lg-6 col-md-6 col-sm-12"
    >
      <h2>
        My Missions
      </h2>
      <div
        className="list-group"
        onKeyDown={[Function]}
      />
    </div>
    <div
      className="pe-lg-5 col-lg-6 col-md-6 col-sm-12"
    >
      <h2>
        My Rockets
      </h2>
      <div
        className="list-group"
        onKeyDown={[Function]}
      />
    </div>
  </div>
</div>
`;

exports[`NavBar page renders correctly 1`] = `
<div>
  <nav
    className="px-5 navbar navbar-expand-lg navbar-light bg-white"
  >
    <a
      className="fs-2 fw-medium navbar-brand"
      href="/"
      onClick={[Function]}
    >
      <img
        alt="planet log"
        height={55}
        src="planet.png"
        width={55}
      />
        
      Space Traveler's Hub
    </a>
    <button
      aria-controls="basic-navbar-nav "
      aria-label="Toggle navigation"
      className="navbar-toggler collapsed"
      onClick={[Function]}
      type="button"
    >
      <span
        className="navbar-toggler-icon"
      />
    </button>
    <div
      aria-expanded={null}
      className="justify-content-end navbar-collapse collapse"
      id="basic-navbar-nav"
    >
      <div
        className="text-primary fs-5 navbar-nav"
        onKeyDown={[Function]}
      >
        <a
          className="text-primary pe-5 nav-link"
          data-rr-ui-event-key={null}
          disabled={false}
          href="/"
          onClick={[Function]}
        >
          Rockets
        </a>
        <a
          className="text-primary nav-link"
          data-rr-ui-event-key={null}
          disabled={false}
          href="/missions"
          onClick={[Function]}
        >
          Missions
        </a>
        <div
          className="d-none d-md-none d-lg-block mt-3 mb-3 vr"
        />
        <a
          className="text-primary nav-link"
          data-rr-ui-event-key={null}
          disabled={false}
          href="/myprofile"
          onClick={[Function]}
        >
          My profile
        </a>
      </div>
    </div>
  </nav>
  <hr />
</div>
`;
